### PR TITLE
Иванов Никита ФИ3. Лабораторная работа № 1.

### PR DESCRIPTION
**Pull request by Atikin-NT**
Реализован плагин, который кидает **warning**, если в названии функции встречается `deprecated`.

### CODE REVIEW

File: clang/test/lab1/ivanov_nikita/dep_matcher.cpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+void foo_deprecated(int a, int b);
+
+// CHECK: warning: Deprecated in function name
+void deprecated(int c);
+
+// CHECK: warning: Deprecated in function name
+void abc();
+
+// CHECK-NOT: warning: Deprecated in function name
```
Comment (m-ly4) [inline clang/test/lab1/ivanov_nikita/dep_matcher.cpp]:
```suggestion
// CHECK-NOT: warning: Deprecated in function name

```

File: clang/labs/lab1/ivanov_nikita/deprecatedMatcher.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        v.TraverseDecl(Context.getTranslationUnitDecl());
+    }
+};
+
+class DeprecatedPlugin : public PluginASTAction {
+protected:
+  std::unique_ptr<ASTConsumer> CreateASTConsumer (
+    CompilerInstance &Compiler, llvm::StringRef InFile) override {
+      return std::unique_ptr<ASTConsumer>(new DepConsumer(Compiler));
```
Comment (m-ly4) [inline clang/labs/lab1/ivanov_nikita/deprecatedMatcher.cpp]: May we create `unique_ptr` without explicit `new` call here?

File: clang/test/lab1/ivanov_nikita/dep_matcher.cpp [hunk 3]
```diff
@@ -0,0 +1,13 @@
+// RUN: %clang_cc1 -load %llvmshlibdir/clangDepMatcher%pluginext -plugin deprecated-match %s 2>&1 | FileCheck %s
+// REQUIRES: plugins
+
+
+void foo_deprecated(int a, int b);
+
+// CHECK: warning: Deprecated in function name
+void deprecated(int c);
+
+// CHECK: warning: Deprecated in function name
```
Comment (m-ly4) [inline clang/test/lab1/ivanov_nikita/dep_matcher.cpp]: Could you please add `help` argument check test?

File: clang/test/lab1/ivanov_nikita/dep_matcher.cpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// RUN: %clang_cc1 -load %llvmshlibdir/clangDepMatcher%pluginext -plugin deprecated-match %s 2>&1 | FileCheck %s
+// REQUIRES: plugins
+
+
+void foo_deprecated(int a, int b);
+
+// CHECK: warning: Deprecated in function name
```
Comment (m-ly4) [inline clang/test/lab1/ivanov_nikita/dep_matcher.cpp]: Could you please move `CHECK` lines closer to functions those checks belong to?

File: clang/labs/lab1/ivanov_nikita/deprecatedMatcher.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  bool ParseArgs(const CompilerInstance &Compiler, 
+                 const std::vector<std::string> &args) override {
+    if (!args.empty() && args[0] == "help")
+      PrintHelp(llvm::errs());
+    
+    return true;
+  }
+
+  void PrintHelp(llvm::raw_ostream& ros) {
+    ros << "Help for DeprecatedPlugin plugin goes here\n";
```
Comment (m-ly4) [inline clang/labs/lab1/ivanov_nikita/deprecatedMatcher.cpp]: Could you please write short but meaningful message here?

File: clang/labs/lab1/ivanov_nikita/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,20 @@
+# set(LLVM_LINK_COMPONENTS Support)
```
Comment (m-ly4) [inline clang/labs/lab1/ivanov_nikita/CMakeLists.txt]: Please remove commented lines if you think they are not required.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by Atikin-NT** [comment]
@m-ly4 I've done, you can check

File: clang/labs/lab1/ivanov_nikita/deprecatedMatcher.cpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+
+using namespace clang;
+
+class DepConsumer : public ASTConsumer {
+  CompilerInstance &Instance;
+
+public:
+    explicit DepConsumer(CompilerInstance &CI): Instance(CI) {}
```
Comment (m-ly4) [inline clang/labs/lab1/ivanov_nikita/deprecatedMatcher.cpp]: Overall, LGTM.
The last remark is: could you please align formatting of your files with the help of clang-format tool?

**Review by m-ly4 — APPROVED** [review]
Thank you.

**Pull request merged**